File: cpp/Lis.cpp
#include "Lis.h"
#include "Komentarz.h"

Lis::Lis(Swiat *swiat, Punkt pole, int turaPowstania)
    : Zwierze(swiat, pole, turaPowstania, sila, inicjatywa, symbol){
    this->symbol = SYMBOL_LISA;
    this->sila = SILA_LISA;
    this->inicjatywa = INICJATYWA_LISA;
    this->pole = pole;
    this->swiat = swiat;
    this->umarl = false;
    this->turaPowstania = turaPowstania;
}

void Lis::akcja(Punkt pole) {
    Punkt nowePole = Punkt(-2, -2);
    while(true) {
        nowePole = losujPunktWZasiegu(pole, ZASIEG_RUCHU_LISA);
        if (nowePole.getY() >= swiat->getY() || nowePole.getY() < 0) continue;
        else if (nowePole.getX() >= swiat->getX() || nowePole.getX() < 0) continue;
        else if(nowePole.getX() == -4 && nowePole.getY() == -4) return;
        else if (!swiat->silniejszyOrganizm(nowePole, INICJATYWA_LISA)) break; //akcja specjalna lisa
        else break;
    }
    if (swiat->czyPoleZajete(nowePole)) {
        swiat->getPole(nowePole)->kolizja(this);
    }
    else {
        swiat->przemiescOrganizm(this, nowePole);
    }
}

string Lis::nazwaOrganizmu() {
    return "Lis";
}

File: java/pl/edu/pg/po/projekt2/zwierzeta/Lis.java
package pl.edu.pg.po.projekt2.zwierzeta;

import pl.edu.pg.po.projekt2.Organizm;
import pl.edu.pg.po.projekt2.Punkt;
import pl.edu.pg.po.projekt2.Swiat;
import pl.edu.pg.po.projekt2.Zwierze;

public class Lis extends Zwierze {
    private static final char symbol = 'L';
    private static final int inicjatywa = 7;

    public Lis(Swiat swiat, Punkt pole, int turaPowstania) {
        super(swiat, pole, turaPowstania, 3, inicjatywa, symbol);
        this.sila = 3;
        this.pole = pole;
        this.swiat = swiat;
        this.umarl = false;
        this.turaPowstania = turaPowstania;
    }

    @Override
    public void akcja(Punkt pole) {
        Punkt nowePole = new Punkt(-2, -2);
        while(true) {
            nowePole = losujPunktWZasiegu(pole, 1);
            if (nowePole.getY() >= swiat.getY() || nowePole.getY() < 0) continue;
            else if (nowePole.getX() >= swiat.getX() || nowePole.getX() < 0) continue;
            else if(nowePole.getX() == -4 && nowePole.getY() == -4) return;
            else if (!swiat.silniejszyOrganizm(nowePole, inicjatywa)) break; //akcja specjalna lisa
            else break;
        }
        if (swiat.czyPoleZajete(nowePole)) {
            swiat.getPole(nowePole).kolizja(this);
        }
        else {
            swiat.przemiescOrganizm(this, nowePole);
        }
    }

    @Override
    public String nazwaOrganizmu() {
        return "Lis";
    }
}
